Added `/core/date-time`

sanderke opened the pull request:
Voorzet voor regel voor datum en tijd

closes #195

sanderke committed:
Added `/core/date-time`

sanderke committed:
markdownlint

TimvdLippe reviewed (dismissed):
Ziet er goed uit! 2 kleine wijzigingen, de rest lijkt mij voldoende om te bespreken in het TO.

sanderke committed:
Apply suggestions from code review

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

TimvdLippe commented on sections/designRules.md:
```diff
@@ -125,6 +125,22 @@ A resource that corresponds to a single conceptual entity is referred to as a [=
+   <dl>
+      <dt>Statement</dt>
+      <dd>
+         <p>All date and time fields in requests and responses MUST be in ISO 8601 format (e.g., <code>YYYY-MM-DD for dates</code>, <code>YYYY-MM-DDTHH:mm:ssZ</code> for timestamps). Fields in responses MUST be in UTC. Fields in requests MUST allow any time offset, which servers SHOULD normalize to (and store in) UTC.</p>
```
```suggestion
         <p>All date and time fields in requests and responses MUST be in [[ISO8601]] format (e.g., <code>YYYY-MM-DD</code> for dates, <code>YYYY-MM-DDTHH:mm:ssZ</code> for timestamps). Fields in responses MUST be in UTC. Servers MUST allow any time offset in requests and SHOULD normalize to (and store in) UTC.</p>
```

TimvdLippe reviewed (commented):
Nog wat kleine fixes. De andere vragen moeten we offline even bespreken wat we daar mee willen.

sanderke committed:
Update sections/designRules.md

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

TimvdLippe committed:
Verwerk feedback op basis van werksessie

TimvdLippe committed:
Vul tekst in bij TODOs

TimvdLippe committed:
Update de bewoording

TimvdLippe committed:
Zet het voorbeeld voor de regels

TimvdLippe committed:
Update bewoording

TimvdLippe committed:
Update bewoording

TimvdLippe committed:
Update zin met time portion

Op basis van review feedback.

TimvdLippe committed:
Voeg noot toe over opslaan van originele timezone

joostfarla commented on sections/designRules.md:
```diff
@@ -125,6 +125,86 @@ A resource that corresponds to a single conceptual entity is referred to as a [=
+      </dd>
+      <dt>Rationale</dt>
+      <dd>
+         <p>Allowing clients to use any timezone offset in requests results in flexibility and less complexity for users. Using UTC in responses results in clarity and removes ambiguity.
```
Hier gaat iets mis met ontbrekende sluit-tags.

TimvdLippe replied:
Closing tags van p elements zijn niet vereist en laten we meestal weg als het niet nodig is voor de leesbaarheid.

joostfarla commented on sections/designRules.md:
```diff
@@ -125,6 +125,86 @@ A resource that corresponds to a single conceptual entity is referred to as a [=
+      <dd>
+         <p>Allowing clients to use any timezone offset in requests results in flexibility and less complexity for users. Using UTC in responses results in clarity and removes ambiguity.
+         <p class="note">This specification does not state rules regarding storage in databases.
+         However, it is recommended to store the originally supplied timezone from the client request in the database, such that later the location of the client can be determined.
```
Dit lijkt me geen algemene recommendation en is niet helemaal wat ik gister bedoelde. Ik zou het formuleren als:
> If the original timezone of a given timestamp value is relevant for users (such as the timezone in which a value is registered), it is recommended to store and publish the timezone details (e.g. the zone offset) as a separate property.

TimvdLippe committed:
Update designRules.md

TimvdLippe commented:
Naar mijn weten is er geen verdere feedback van TO-leden tot nu toe, dus ik heb het label "Ter goedkeuring" toegevoegd voor bespreking op het TO van 2 december.

Het voorstel vanuit TO-leden is om tijdens het TO deze wijziging vast te stellen, maar daarna nog een consultatie te doen specifiek bij het Kennisplatform API's. Op basis van de feedback kunnen we dan in het eerste TO van 2026 de definitieve vaststelling in het TO doen en deze wijziging doorvoeren.

TimvdLippe committed:
Voeg linter regels toe voor datum en tijd

TimvdLippe commented:
Enkele linter regels geschreven. Hierbij de resultaten:

```
Statistics for rule nlgov:date-time-ensure-timezone:

Analyzed: 160
Passing: 160
Percentage: 100%

Statistics for rule nlgov:time-without-timezone:

Analyzed: 160
Passing: 156
Percentage: 97%

Statistics for rule nlgov:specify-format-for-date-and-time:

Analyzed: 160
Passing: 140
Percentage: 87%

Statistics for rule nlgov:use-date-instead-of-datetime:

Analyzed: 160
Passing: 152
Percentage: 95%
```

Interessante bevindingen:
1. De heuristieken werken best wel redelijk. Het is lastig om te bepalen wanneer iets een datum had moeten zijn, maar door op basis van tekst wat aannames te doen lijken er de correcte gevallen uitgepikt te worden.
2. We zeggen iets over fields, maar moeten we ook iets zeggen over parameters? Dat wordt nog best vaak gebruikt.
3. Moeten we enkel `time-local` toestaan of is `time` soms ook toegestaan? E.g. met tijdzone, maar dan zonder datum.

Voor dat laatste punt, wat denken we van dit soort gevallen? Volgens mij hebben ze hier `time-local` bedoeld ipv `time` (zie beschrijving op https://spec.openapis.org/registry/format/):

```
{
  "name": "tijdstipWeging[in]",
  "in": "query",
  "description": "Matches any value from a comma-separated list: val1,val2,valN.",
  "schema": {
    "type": "array",
    "items": {
      "type": "string",
      "format": "time"
    }
  },
  "explode": false,
  "style": "form"
},
```

TimvdLippe commented:
Deze wijziging is goedgekeurd bij het TO 2025-12-02 voor consultatie bij Kennisplatform API's. Dat houdt in dat we het hoofdstuk "Date & Time" als geheel consulteren.

Tevens was er nog een behoefte aan een aparte regel dat als er zowel een tijd als datum veld apart in een response zitten, die moeten worden samengevoegd tot 1 veld.

Daarnaast nog een noot toevoegen dat er goed moet worden nagedacht als er enkel `time` wordt gebruikt, maar met een terugkerend patroon. In dat geval kan het in de tijdzone overgang vallen, waar de tijdzone van belang is. In dat geval is een datetime waarschijnlijk geschikter.

joostfarla committed:
Storage recommendations (#298)

TimvdLippe committed:
Merge branch 'develop' into date-time

TimvdLippe committed:
Remove example formats for ISO8601

To avoid confusion whether other variations of these
accepted formats are allowed.

TimvdLippe committed:
Gebruik meest recente versie van ISO8601

TimvdLippe committed:
Response in UTC als SHOULD

bwbroersma commented:
It's not ideal to require the intersection of ISO 8601 and some RFCs, since the ISO standard is not open.

TimvdLippe committed:
Leg het verschil uit tussen RFC9557 en ISO8601

Hiermee voorkomen we het probleem dat het vereist is om
ISO8601 te kopen (dus blijft het een open standaard).

TimvdLippe commented:
Samen met Benjamin dat probleem verholpen door uit te leggen wat het verschil is tussen RFC9557 en ISO8601.

TimvdLippe committed:
Fix MUST notatie

TimvdLippe committed:
Fix ISO8601 formaat voor time

`time` bestaat niet in de ABNF, dit moet `partial-time` zijn

TimvdLippe committed:
Fix typo

KlaasBaarssen commented:
Het punt is duidelijk verbeterd, maar 1 zorgpunt blijft. Dit gaat over interpretatie van een gegeven.
In de tekst staat het al: geboortedatum. Het is mij niet duidelijk hoe voorkomen wordt of een datetime wel of niet geconverteerd mag worden naar lokale tijd. In het voorbeeld van geboortedatum <code>2025-03-20T00:00:00+01:00</code>, timezone conversion would result in <code>2025-03-19T23:00:00Z</code>
Voor een vergadering is het wel handig dat een date-time wordt opgezet naar lokale tijdzone, dit om te zorgen dat de medewerker in de andere tijdzone er op het juiste tijdstip is. 
Na het lezen over de verschillende types is het voor mij niet duidelijk welk type ik moet inzetten om conversie aan te moedigen of juist te ontmoedigen.

TimvdLippe commented:
[/core/date-time/date-omit-time-portion](https://logius-standaarden.github.io/Publicatie-Preview/API-Design-Rules/date-time/#/core/date-time/date-omit-time-portion) schrijft voor dat als het tijdzone-component niet relevant is, er een `date` moet worden gebruikt in plaats van `datetime`. Verduidelijkt dat de situatie?

In het geval van een geboortedatum is de tijdzone dus niet relevant, dus moet het `date` zijn. Voor een vergadering is de tijdzone wel relevant, dus is het `datetime`.

Voor verdere achtergrond, zie ook https://apiux.com/2013/03/20/5-laws-api-dates-and-times/ (law 5) waar dit op gebaseerd is. Deze blogpost is de basis voor alle regels in het "Date & Time" hoofdstuk en is internationaal erkend als de facto standaard hoe om te gaan met datum en tijd.

KlaasBaarssen commented:
> [/core/date-time/date-omit-time-portion](https://logius-standaarden.github.io/Publicatie-Preview/API-Design-Rules/date-time/#/core/date-time/date-omit-time-portion) schrijft voor dat als het tijdzone-component niet relevant is, er een `date` moet worden gebruikt in plaats van `datetime`. Verduidelijkt dat de situatie?
> 
> In het geval van een geboortedatum is de tijdzone dus niet relevant, dus moet het `date` zijn. Voor een vergadering is de tijdzone wel relevant, dus is het `datetime`.
> 
> Voor verdere achtergrond, zie ook https://apiux.com/2013/03/20/5-laws-api-dates-and-times/ (law 5) waar dit op gebaseerd is. Deze blogpost is de basis voor alle regels in het "Date & Time" hoofdstuk en is internationaal erkend als de facto standaard hoe om te gaan met datum en tijd.

Geboortedatum is vaak een date, maar in een geboorteakte vaak een datetime, omdat daar tijdstip van geboorte bij staat. Het gekke is dat als je die datum wel converteert naar lokale tijd het technisch klopt. Stel in NL geboren om 14:00+2 dan is het voor UK 12:00+0 geboren. Klopt technisch, maar wenselijk.....
Daarnaast heb je bijv. rondom een zoekfunctie met datum+tijd een uitdaging. Stel je server draait in US en je bent in NL en geeft aan dat je alle reserveringen wilt zien voor dag X tussen 14:00 en 15:00. Als dit met tijdzone wordt gestuurd van NL (+ 1 of + 2), moet de server dit omzetten naar US tijdzone (meerdere opties) en krijg je de reserveringen terug die tussen 14:00 en 15:00 in NL-tijd, maar US-tijd paar uur eerder. Het kan dat je dit wilt, maar ook niet afhankelijk of boekingstijden NL of US tijden zijn. En deze onduidelijkheid wordt niet weggenomen met deze ADR.

TimvdLippe committed:
Maak tekst over timezone offset uitgebreider

Voeg ook voorbeelden toe wanneer wat te doen
en ook hoe om te gaan met een relevante
timezone offset.

TimvdLippe commented:
Het punt van Klaas met hem besproken en zojuist een update gepusht waarbij het gedeelte van de noot over "als de timezone offset relevant is" verplaatst naar de statement. Tevens 2 voorbeelden toegevoegd voor die scenario's en hoe daar dan mee om moet worden gegaan.

TimvdLippe commented:
Deze wijziging is goedgekeurd op het TO API van 2026-03-03.

BvanRaaij commented:
Mooi resultaat!

TimvdLippe committed:
Merge branch 'develop' into date-time
